image/image.go

```diff
@@ -0,0 +1,386 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package image
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/arch"
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/boot"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/partition"
+	"github.com/snapcore/snapd/progress"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/squashfs"
+	"github.com/snapcore/snapd/store"
+)
+
+type Options struct {
+	Snaps            []string
+	Rootdir          string
+	Channel          string
+	ModelAssertionFn string
+	GadgetUnpackDir  string
+}
+
+func Prepare(opts *Options) error {
+	if err := downloadUnpackGadget(opts); err != nil {
+		return err
+	}
+
+	return bootstrapToRootdir(opts)
+}
+
+func decodeModelAssertion(fn string) (*asserts.Model, error) {
+	rawAssert, err := ioutil.ReadFile(fn)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read model assertion: %s", err)
+	}
+
+	ass, err := asserts.Decode(rawAssert)
+	if err != nil {
+		return nil, fmt.Errorf("cannot decode model assertion %q: %s", fn, err)
+	}
+	return ass.(*asserts.Model), nil
+}
+
+func downloadUnpackGadget(opts *Options) error {
+	model, err := decodeModelAssertion(opts.ModelAssertionFn)
+	if err != nil {
+		return err
+	}
+	if err := os.MkdirAll(opts.GadgetUnpackDir, 0755); err != nil {
+		return fmt.Errorf("cannot create gadget unpack dir %q: %s", opts.GadgetUnpackDir, err)
+	}
+
+	dlOpts := &downloadOptions{
+		TargetDir:    opts.GadgetUnpackDir,
+		Channel:      opts.Channel,
+		StoreID:      model.Store(),
+		Architecture: model.Architecture(),
+	}
+	snapFn, _, err := acquireSnap(model.Gadget(), dlOpts)
+	if err != nil {
+		return err
+	}
+	// FIXME: jumping through layers here, we need to make
+	//        unpack part of the container interface (again)
+	snap := squashfs.New(snapFn)
+	return snap.Unpack("*", opts.GadgetUnpackDir)
+}
+
+func acquireSnap(snapName string, dlOpts *downloadOptions) (string, *snap.Info, error) {
+	if osutil.FileExists(snapName) {
+		return copyLocalSnapFile(snapName, dlOpts.TargetDir)
+	}
+
+	return downloadSnapWithSideInfo(snapName, dlOpts)
+}
+
+func bootstrapToRootdir(opts *Options) error {
+	if opts.Rootdir != "" {
+		dirs.SetRootDir(opts.Rootdir)
+		defer dirs.SetRootDir("/")
```

Comment:
How hard would it be to not have this?  In other words, do we depend on logic elsewhere looking at their usual paths under the root?

I ask because doing this means we cannot ever call this function concurrently with anything else, as it'd break down typical work.